### PR TITLE
Adds CNN Trainer

### LINKED ISSUE #203: Re usable Components and CNN Trainer
Hey all, I'm new to Bolts and a bit to lightning, but wokred with PyTorch. This project is super awesome, completely in snyc with my thoughts.

## 🚀 Feature

Torchivision supports lot of pre-trained CNN models. One of the common tasks in fine tuning these CNNs. (Transfer learning) This issue was open with lightning team as well I guess. Can we create a trainer for this purpose that allows easy fine tuning or even raw training just by using `pretrained = False`

The only layer that differs in the last Linear layer which being `nn.Linear(hidden_dim, num_classes)`

### Motivation

Such pre-trained CNNs are used as Backbones for Object detection tasks as well. E.g. you can fine tune a CNN on say animals such as tiger, lion and then use these CNN weights as backbone for F-RCNN .

This is a key feature, lightning supports multi GPU training and TPU training out of the box for all such operations.

### Pitch

One of my earlier attempts to do the same with pytorch was [here](https://github.com/oke-aditya/pytorch_cnn_trainer). I could successfully tune and train all the  PyTorch models and almost all [PyTorch Image Models](https://github.com/rwightman/pytorch-image-models). It had simple API which syncs with #200.

`model = model_factory.create_torchvision_model("resnet18", num_classes=10, pretrained=True,)`

All these fine-tunable backbones are compatible with F-RCNN also, they can simply be passed to torchvision FRCNN.

### Alternatives

I'm unsure if my backbone extraction for CNN is efficient and good. It assumes we have `nn.AdaptiveAvgPool2d()` followed by Linear only in classification tasks, hence simply extracts all layers before this linear. I held to this assumption since all torchvision and [PyTorch Image Models](https://github.com/rwightman/pytorch-image-models) follow this. But I guess we need to assume something.

### Additional Context

What this feature allows is simple fine tuner to CNNs directly supported with all lightning perks and bonuses.
A common backbone API, re-usable to all vision models such as FRCNN, Mask-RCNN, RetinaNet (in next torchvision release).
I'm unsure still if this can be extensible and maintainable.

I can probably work on this feature, I would need a bit help as I'm new to this repo and lightning.

### PR DESCRIPTION
**Pull request by oke-aditya**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Fixes #203 

Adds a generic CNN Trainer that works for torchvision models.

## PR review

I need a bit help to write tests and with the `run_cli()` method. I'm not sure if they are right.
I have tested this code and it works fine in my repo. It is [here](https://github.com/oke-aditya/pytorch_cnn_trainer)

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@oke-aditya how is it going here, still wip?

**Comment by oke-aditya** [comment]
Hey @Borda,

Actually, the code is complete. I need to add tests as well though.

@justusschock said this might need more discussion hence I kept as WIP.

We can discuss on this PR, I guess it is good feature addition though.

This can be used to solve #340 too.

**Comment by Borda** [comment]
let's close this one for now... :]
